### PR TITLE
Implement NGNX token

### PR DESCRIPTION
**Pull request by njokuScript**
Implement NGNX token

- Implement standard ERC20 function
- Implement gasless approvals
- Implement system caging
- Implement minting and burning functionality
- Make contract upgradeable

### CODE REVIEW

File: contracts/ngnx.sol [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender));
+        _grantRole(MINTER_ROLE, account);
+    }
+
+    /**
+     * @dev Mints a new token
+     * @param account address to send the minted tokens to
+     * @param amount amount of tokens to mint
+     */
+    function mint(address account, uint amount) external returns (bool) {
```
Comment (iamnotstatic) [inline contracts/ngnx.sol]: Can we use solidity custom error, as it is more optimal and saves gas.

File: contracts/ngnx.sol [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        address tokenAddress,
+        address account
+    ) public virtual {
+        IERC20 token = IERC20(tokenAddress);
+        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender));
+        uint256 tokenBalance = token.balanceOf(address(this));
+        token.transfer(account, tokenBalance);
+    }
+
+    function cage() public {
```
Comment (iamnotstatic) [inline contracts/ngnx.sol]: For clarification, once it's 0, it can never be set to 1 right ?

File: contracts/ngnx.sol [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        return true;
+    }
+
+    /**
+     * @dev Burns a  token
+     * @param account address to burn tokens from
+     * @param amount amount of tokens to burn
+     */
+    function burn(address account, uint amount) external returns (bool) {
+        require(live == 1, "NGNX/not-live");
```
Comment (MeisterJustice) [inline contracts/ngnx.sol]: Curious to know why you're checking for `require(live == 1, "NGNX/not-live");` when you already set live to be 1 in the initializer?

File: contracts/ngnx.sol [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        if (
+            account != msg.sender &&
+            allowance(msg.sender, account) != type(uint).max
+        ) {
+            require(
+                allowance(msg.sender, account) >= amount,
+                "NGNx/insufficient-allowance"
+            );
+            decreaseAllowance(account, amount);
```
Comment (MeisterJustice) [inline contracts/ngnx.sol]: Where is this `allowance` gotten from and what's it doing? Also, why do we user `msg.sender` and `account` interchangeably? `msg.sender` should be enough

**Review by MeisterJustice — CHANGES_REQUESTED** [review]

File: contracts/ngnx.sol (hunk 3 above)
Comment (MeisterJustice) [reply contracts/ngnx.sol]: The question is, wouldn't live be always equal to 1?

Comment (njokuScript) [reply contracts/ngnx.sol]: live can be 0. when we want to stop the system

we check for the live state when we want to carry out minting or burning.

File: contracts/ngnx.sol (hunk 4 above)
Comment (njokuScript) [reply contracts/ngnx.sol]: The allowance is gotten from `ERC20Upgradeable`

That function accepts 2 parameters `owner` & `spender`. ref: https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-allowance-address-address-

File: contracts/ngnx.sol (hunk 2 above)
Comment (njokuScript) [reply contracts/ngnx.sol]: Yeah. this is used in a case of a bug or hack.

File: contracts/ngnx.sol [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+     */
+    function burn(address account, uint amount) external returns (bool) {
+        if (live != 1) {
+            revert NotLive("NGNX/not-live");
+        }
+        if (
+            account != msg.sender &&
+            allowance(msg.sender, account) != type(uint).max
+        ) {
+            require(
```
Comment (iamnotstatic) [inline contracts/ngnx.sol]: For the sake of consistency you can also add the custom error here

Comment (njokuScript) [reply contracts/ngnx.sol]: sorry, missed it

**Review by iamnotstatic — APPROVED** [review]
LGTM 🚀

File: contracts/ngnx.sol (hunk 3 above)
Comment (MeisterJustice) [reply contracts/ngnx.sol]: There's no method to update the value of live. As it stands, live will always be 1 when deployed, unless the contract is upgraded later

**Pull request merged**